### ISSUE TITLE
Parsing DOT string with formatting in it

### ISSUE THREAD

**Issue by s4m0r4m4** [visjs/vis-network#505]
Hello, I am using `vis.parseDOTNetwork(sTemp);` to parse this DOT-style string:

```
digraph Tree {
node [shape=box, style="filled, rounded", color="black", fontname=helvetica] ;
edge [fontname=helvetica] ;
0 [label=<Intensity &lt; 120 | Height (px) &le; 2.5<br/>gini = 0.158<br/>samples = 8946<br/>value = [8175, 771]<br/>class = No Defect>, fillcolor="#e78d4c"] ;
1 [label=<gini = 0.0<br/>samples = 7508<br/>value = [7508, 0]<br/>class = No Defect>, fillcolor="#e58139"] ;
0 -> 1 [labeldistance=2.5, labelangle=45, headlabel="True"] ;
2 [label=<Intensity &lt; 120 | Width (px) &le; 20.5<br/>gini = 0.497<br/>samples = 1438<br/>value = [667, 771]<br/>class = Defect>, fillcolor="#e4f2fb"] ;
0 -> 2 [labeldistance=2.5, labelangle=-45, headlabel="False"] ;
3 [label=<Intensity &lt; 120 | Area (px) &le; 22.5<br/>gini = 0.431<br/>samples = 875<br/>value = [275, 600]<br/>class = Defect>, fillcolor="#94caf1"] ;
2 -> 3 ;
4 [label=<gini = 0.371<br/>samples = 191<br/>value = [144, 47]<br/>class = No Defect>, fillcolor="#edaa7a"] ;
3 -> 4 ;
5 [label=<gini = 0.31<br/>samples = 684<br/>value = [131, 553]<br/>class = Defect>, fillcolor="#68b4eb"] ;
3 -> 5 ;
6 [label=<Intensity &lt; 120 | Height (px) &le; 4.5<br/>gini = 0.423<br/>samples = 563<br/>value = [392, 171]<br/>class = No Defect>, fillcolor="#f0b88f"] ;
2 -> 6 ;
7 [label=<gini = 0.491<br/>samples = 306<br/>value = [174, 132]<br/>class = No Defect>, fillcolor="#f9e1cf"] ;
6 -> 7 ;
8 [label=<gini = 0.257<br/>samples = 257<br/>value = [218, 39]<br/>class = No Defect>, fillcolor="#ea985c"] ;
6 -> 8 ;
}
```

The DOT string was created using scikit-learn's `sklearn.tree.export_graphviz(...)` (see https://scikit-learn.org/stable/modules/generated/sklearn.tree.export_graphviz.html#sklearn.tree.export_graphviz). I have verified that this is a valid DOT string using [GraphViz Online](https://dreampuf.github.io/GraphvizOnline/#digraph%20Tree%20%7B%0D%0Anode%20%5Bshape%3Dbox%2C%20style%3D%22filled%2C%20rounded%22%2C%20color%3D%22black%22%2C%20fontname%3Dhelvetica%5D%20%3B%0D%0Aedge%20%5Bfontname%3Dhelvetica%5D%20%3B%0D%0A0%20%5Blabel%3D%3CIntensity%20%26lt%3B%20120%20%7C%20Height%20(px)%20%26le%3B%202.5%3Cbr%2F%3Egini%20%3D%200.158%3Cbr%2F%3Esamples%20%3D%208946%3Cbr%2F%3Evalue%20%3D%20%5B8175%2C%20771%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23e78d4c%22%5D%20%3B%0D%0A1%20%5Blabel%3D%3Cgini%20%3D%200.0%3Cbr%2F%3Esamples%20%3D%207508%3Cbr%2F%3Evalue%20%3D%20%5B7508%2C%200%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23e58139%22%5D%20%3B%0D%0A0%20-%3E%201%20%5Blabeldistance%3D2.5%2C%20labelangle%3D45%2C%20headlabel%3D%22True%22%5D%20%3B%0D%0A2%20%5Blabel%3D%3CIntensity%20%26lt%3B%20120%20%7C%20Width%20(px)%20%26le%3B%2020.5%3Cbr%2F%3Egini%20%3D%200.497%3Cbr%2F%3Esamples%20%3D%201438%3Cbr%2F%3Evalue%20%3D%20%5B667%2C%20771%5D%3Cbr%2F%3Eclass%20%3D%20Defect%3E%2C%20fillcolor%3D%22%23e4f2fb%22%5D%20%3B%0D%0A0%20-%3E%202%20%5Blabeldistance%3D2.5%2C%20labelangle%3D-45%2C%20headlabel%3D%22False%22%5D%20%3B%0D%0A3%20%5Blabel%3D%3CIntensity%20%26lt%3B%20120%20%7C%20Area%20(px)%20%26le%3B%2022.5%3Cbr%2F%3Egini%20%3D%200.431%3Cbr%2F%3Esamples%20%3D%20875%3Cbr%2F%3Evalue%20%3D%20%5B275%2C%20600%5D%3Cbr%2F%3Eclass%20%3D%20Defect%3E%2C%20fillcolor%3D%22%2394caf1%22%5D%20%3B%0D%0A2%20-%3E%203%20%3B%0D%0A4%20%5Blabel%3D%3Cgini%20%3D%200.371%3Cbr%2F%3Esamples%20%3D%20191%3Cbr%2F%3Evalue%20%3D%20%5B144%2C%2047%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23edaa7a%22%5D%20%3B%0D%0A3%20-%3E%204%20%3B%0D%0A5%20%5Blabel%3D%3Cgini%20%3D%200.31%3Cbr%2F%3Esamples%20%3D%20684%3Cbr%2F%3Evalue%20%3D%20%5B131%2C%20553%5D%3Cbr%2F%3Eclass%20%3D%20Defect%3E%2C%20fillcolor%3D%22%2368b4eb%22%5D%20%3B%0D%0A3%20-%3E%205%20%3B%0D%0A6%20%5Blabel%3D%3CIntensity%20%26lt%3B%20120%20%7C%20Height%20(px)%20%26le%3B%204.5%3Cbr%2F%3Egini%20%3D%200.423%3Cbr%2F%3Esamples%20%3D%20563%3Cbr%2F%3Evalue%20%3D%20%5B392%2C%20171%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23f0b88f%22%5D%20%3B%0D%0A2%20-%3E%206%20%3B%0D%0A7%20%5Blabel%3D%3Cgini%20%3D%200.491%3Cbr%2F%3Esamples%20%3D%20306%3Cbr%2F%3Evalue%20%3D%20%5B174%2C%20132%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23f9e1cf%22%5D%20%3B%0D%0A6%20-%3E%207%20%3B%0D%0A8%20%5Blabel%3D%3Cgini%20%3D%200.257%3Cbr%2F%3Esamples%20%3D%20257%3Cbr%2F%3Evalue%20%3D%20%5B218%2C%2039%5D%3Cbr%2F%3Eclass%20%3D%20No%20Defect%3E%2C%20fillcolor%3D%22%23ea985c%22%5D%20%3B%0D%0A6%20-%3E%208%20%3B%0D%0A%7D).

When I call `parseDOTNetwork` , it throws the following error: "`Syntax error in part "<Intensity &lt; 120 | Heigh..."`" from line 7283 of vis-network.js.

It seems like vis-network is not set up to parse more complex labels with HTML formatting, but I could not find anything in the docs describing the limitations or workarounds. What do I need to do to be able to parse this DOT string?

Here are my versions from package-lock.json:
```
"vis-data": {
			"version": "6.2.1",
			"resolved": "https://registry.npmjs.org/vis-data/-/vis-data-6.2.1.tgz",
			"integrity": "sha512-y80h+jJ6gQXfeWqDdts9UTrLubggx99Q0dJionQF/XxHIbcxNZ+2cuwv4y+bIapVF9nW0DyE08EcMMlicAlSWg==",
			"requires": {
				"vis-util": "^1.1.0"
			}
		},
"vis-network": {
			"version": "7.3.5",
			"resolved": "https://registry.npmjs.org/vis-network/-/vis-network-7.3.5.tgz",
			"integrity": "sha512-i9Ow9qqYGhXwcTEetbKCHBTzYvjsag5xM+Zl009ajjZyAvMLqu3Gljk/VpcAn2WqQHZAz0g/Cg3+xfALxYwLGA=="
		},
"vis-util": {
			"version": "1.1.8",
			"resolved": "https://registry.npmjs.org/vis-util/-/vis-util-1.1.8.tgz",
			"integrity": "sha512-gX461BUrYnHfDpMalvQdQ26SG/DplIvsK7u5KUCWEfPpcnTXm22FtQH5a+GIbDBcePe1sm578mpbWwTbCgRVww==",
			"requires": {
				"moment": "^2.24.0",
				"vis-uuid": "^1.1.3"
			}
		},
```

**Comment by s4m0r4m4** [comment]
Hello, I'm just following up on this - is there any restriction on what version of DOT I can use? Or are certain DOT features not supported? I don't necessarily need the code updated, I just need to know what it can support so that I can work around the features that are not supported.

**Comment by s4m0r4m4** [comment]
It also looks like it does not support subgraphs:

```
digraph Tree { fontsize=25;
subgraph cluster_1 {
        node [color=black fontname=helvetica shape=box style="filled, rounded"]
        edge [fontname=helvetica labelangel=45 labeldistance=2.5]
        V1_0 [label="Intensity Threshold: < 170
Defect (#): 256
Okay (#): 2294
Split Gain: 0.081" fillcolor="#199750"]
        V1_0 -> V1_1 [xlabel="Area >= 967"]
        V1_1 [label="Intensity Threshold: > 50
Defect (#): 149
Okay (#): 30
Split Gain: 0.068" fillcolor="#ea5739"]
label = "View 1"
}
subgraph cluster_2 {
        node [color=black fontname=helvetica shape=box style="filled, rounded"]
        edge [fontname=helvetica labelangel=45 labeldistance=2.5]
        V2_0 [label="Intensity Threshold: > 60
Defect (#): 1049
Okay (#): 2067
Split Gain: 0.111" fillcolor="#b9e176"]
        V2_0 -> V2_1 [xlabel="Std >= 22.2"]
        V2_1 [label="Intensity Threshold: > 160
Defect (#): 818
Okay (#): 532
Split Gain: 0.048" fillcolor="#fedc88"]
label = "View 2"
}
}
```

**Comment by s4m0r4m4** [comment]
Another gap: It also does not render "xlabels" on edges nor nodes.

Looking at the old vis.js (which I believe [this demo](https://dreampuf.github.io/GraphvizOnline/#digraph%20cluster_0%20%7B%0D%0Anode%20%5Bcolor%3Dblack%20fontname%3Dhelvetica%20shape%3Dbox%20style%3D%22filled%2C%20rounded%22%5D%0D%0Aedge%20%5Bfontname%3Dhelvetica%20labelangel%3D45%20labeldistance%3D2.5%5D%0D%0AV0_0%20%5Blabel%3D%22Threshold%3A%20%3C120-Defect%20(%23)%3A%20173-Okay%20(%23)%3A%201858-Split%20Gain%3A%200.077%22%20fillcolor%3D%22%2315904c%22%5D%0D%0AV0_0%20-%3E%20V0_1%20%5Bxlabel%3D%22Height%20%3E%3D%203%22%5D%0D%0AV0_1%20%5Blabel%3D%22Threshold%3A%20%3C120-Defect%20(%23)%3A%20173-Okay%20(%23)%3A%20149-Split%20Gain%3A%200.084%22%20fillcolor%3D%22%23fff3ac%22%5D%0D%0AV0_1%20-%3E%20V0_2%20%5Bxlabel%3D%22Width%20%3E%3D%2021%22%5D%0D%0AV0_2%20%5Blabel%3D%22Threshold%3A%20%3C130-Defect%20(%23)%3A%2036-Okay%20(%23)%3A%2091-Split%20Gain%3A%200.047%22%20fillcolor%3D%22%239bd469%22%5D%0D%0AV0_3%20%5Blabel%3D%22Class%3A%20Defect-Defect%20(%23)%3A%207-Okay%20(%23)%3A%201-%22%20fillcolor%3D%22%23de402e%22%5D%0D%0AV0_4%20%5Blabel%3D%22Class%3A%20Okay-Defect%20(%23)%3A%2029-Okay%20(%23)%3A%2090-%22%20fillcolor%3D%22%2382c966%22%5D%0D%0AV0_1%20-%3E%20V0_5%20%5Blabel%3D%22Width%20%3C%2021%22%5D%0D%0AV0_5%20%5Blabel%3D%22Threshold%3A%20%3C130-Defect%20(%23)%3A%20137-Okay%20(%23)%3A%2058-Split%20Gain%3A%200.105%22%20fillcolor%3D%22%23fdad60%22%5D%0D%0AV0_5%20-%3E%20V0_6%20%5Bxlabel%3D%22Area%20%3E%3D%2037%22%5D%0D%0AV0_6%20%5Blabel%3D%22Class%3A%20Defect-Defect%20(%23)%3A%20113-Okay%20(%23)%3A%2018-%22%20fillcolor%3D%22%23e24731%22%5D%0D%0AV0_5%20-%3E%20V0_7%20%5Bxlabel%3D%22Area%20%3C%2037%22%5D%0D%0AV0_7%20%5Bxlabel%3D%22Class%3A%20Okay-Defect%20(%23)%3A%2024-Okay%20(%23)%3A%2040-%22%20fillcolor%3D%22%23cbe982%22%5D%0D%0AV0_0%20-%3E%20V0_8%20%5Bxlabel%3D%22Height%20%3C%203%22%5D%0D%0AV0_8%20%5Blabel%3D%22Class%3A%20Okay-Defect%20(%23)%3A%200-Okay%20(%23)%3A%201709-%22%20fillcolor%3D%22%23006837%22%5D%0D%0AV0_2%20-%3E%20V0_4%20%5Bxlabel%3D%22Height%20%3C%2010%22%5D%0D%0AV0_2%20-%3E%20V0_3%20%5Bxlabel%3D%22Height%20%3E%2010%22%5D%0D%0A%7D) runs on), it seems to render that stuff no problem. Maybe there is some old code from there that you guys can use?

**Comment by s4m0r4m4** [comment]
Ping - just checking in on this

**Comment by Thomaash** [comment]
Hi @s4m0r4m4,

the DOT parser is old, definitely incomplete, not exactly specs compliant even where implemented and yeah, I can't rule out that it is indeed broken.

Would you (@s4m0r4m4 or whoever is reading this) be willing to work on this? All the logic is neatly confined (not a single import/require statement) within https://github.com/visjs/vis-network/blob/master/lib/network/dotparser.js so not much insight into the library is needed (DOT string is the input and the output is directly supplied to `network.setData({ nodes, edges })`).

**Comment by s4m0r4m4** [comment]
Hi Thomassh, thanks for the message. I will have to see when this rises to the priority queue for me, it does sound nicely compartmentalized. Are you open to replacing most of that functionality with a separate lib (e.g. https://github.com/anvaka/dotparser/ or something similar)?

**Comment by HighVoltageCoder** [comment]
Hi, 
I've also noticed problem with double quotes parsing. Example:
`27 [Cmd="\"C:\Program Files (x86)\Microsoft Office\Root\Office16\SDXHelper.exe\" -Embedding"].
`
I looked into the source code and it looks like it does handle those cases when there are double quotes. Parser takes the first (") and thinks that the second one is the end of string and the rest of the string shouldn't be there.
Do you know any workaround ? :)

**Comment by Huyston** [comment]
It also doesn't like comments. If I remove comments (//) it works.

**Comment by vorburger** [comment]
I've run into this as well while trying to render Graphviz from (my) [Enola.dev RDF Graph Generation](https://docs.enola.dev/models/example.org/graph/) with Vis.js. (In my specific case, after finding and reading this issue, I'm going to "skip" this Dotparser and just "directly" turn my Things data structure into what Vis.js "natively" needs.)

PS: If any maintainers are reading this, I would duplicate #1957 to this issue.
